skorch/helper.py

```diff
@@ -508,3 +509,83 @@ def describe_signature(self, df):
             )
 
         return signature
+
+
+class AccelerateMixin:
+    """Mixin class to add support for huggingface's accelerate
+
+    *Experimental*
+
+    Examples
+    --------
+    >>> from skorch import NeuralNetClassifier
+    >>> from skorch.helper import AccelerateMixin
+    >>> from accelerate import Accelerator
+
+    >>> class AcceleratedNet(AccelerateMixin, NeuralNetClassifier):
+    >>>     pass
+
+    >>> accelerator = Accelerator(...)
+    >>> net = AcceleratedNet(
+    ...     MyModule,
+    ...     accelerator=accelerator,
+    ...     device=None,
+    ...     callbacks__print_log__sink=accelerator.print)
+    >>> net.fit(X, y)
+
+    """
+    def __init__(self, *args, accelerator, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.accelerator = accelerator
+
+        if self.accelerator.device_placement and (self.device is not None):
+            raise ValueError(
+                "When device placement is performed be accelerator, set device=None")
+
+    def _initialize_criterion(self, *args, **kwargs):
+        super()._initialize_criterion(*args, **kwargs)
+
+        with self._current_init_context('criterion'):
+            for name in self._criteria:
+                criterion = getattr(self, name + '_')
+                if isinstance(criterion, torch.nn.Module):
+                    self.accelerator.prepare(criterion)
+
+            return self
+
+    def _initialize_module(self, *args, **kwargs):
+        super()._initialize_module(*args, **kwargs)
+
+        with self._current_init_context('module'):
+            for name in self._modules:
+                module = getattr(self, name + '_')
+                if isinstance(module, torch.nn.Module):
+                    self.accelerator.prepare(module)
+
+        return self
+
+    def _initialize_optimizer(self, *args, **kwargs):
+        super()._initialize_optimizer(*args, **kwargs)
+
+        with self._current_init_context('optimizer'):
+            for name in self._optimizers:
+                optimizer = getattr(self, name + '_')
+                if isinstance(optimizer, torch.optim.Optimizer):
+                    self.accelerator.prepare(optimizer)
+        return self
+
+    def train_step_single(self, batch, **fit_params):
+        self._set_training(True)
+        Xi, yi = unpack_data(batch)
+        y_pred = self.infer(Xi, **fit_params)
+        loss = self.get_loss(y_pred, yi, X=Xi, training=True)
+        self.accelerator.backward(loss)
```

Comment:
The [accelerator docs](https://huggingface.co/docs/accelerate/index.html) says it better to let `accelerate` handle device transfers. Is it possible to do the same in `skorch`?

Reply:
I thought this is happening when `device_placement=True`, which is the default? That's why I check that attribute during initialization and require `device=None` in that case, i.e. that skorch does not bother with moving to the desired device. Or do you mean something else?

Reply:
Ah I see, I missed the `device=None`.